aws-lc-rs/src/aead/aead_ctx.rs

```diff
@@ -20,58 +39,223 @@ use aws_lc::{
+    ) -> Result<LcPtr<*mut EVP_AEAD_CTX>, Unspecified> {
+        if AES_256_KEY_LEN != key_bytes.len() {
+            return Err(Unspecified);
+        }
+        AeadCtx::build_context(aead, key_bytes, tag_len, direction)
+    }
+
     fn build_context(
         aead_fn: unsafe extern "C" fn() -> *const aws_lc::evp_aead_st,
         key_bytes: &[u8],
+        tag_len: usize,
+        direction: Option<AeadDirection>,
     ) -> Result<LcPtr<*mut EVP_AEAD_CTX>, Unspecified> {
         let aead = unsafe { aead_fn() };
 
-        let aead_ctx = unsafe {
-            LcPtr::new(EVP_AEAD_CTX_new(
-                aead,
-                key_bytes.as_ptr().cast(),
-                key_bytes.len(),
-                TAG_LEN,
-            ))?
-        };
+        // We are performing the allocation ourselves as EVP_AEAD_CTX_new will call EVP_AEAD_CTX_init by default
+        // and this avoid having to zero and reinitalize again if we need to set an explicit direction.
+        let aead_ctx: LcPtr<*mut EVP_AEAD_CTX> =
+            LcPtr::new(unsafe { OPENSSL_malloc(size_of::<EVP_AEAD_CTX>()) }.cast())?;
+
+        unsafe { EVP_AEAD_CTX_zero(*aead_ctx) };
 
+        if 1 != match direction {
+            Some(direction) => unsafe {
+                EVP_AEAD_CTX_init_with_direction(
+                    *aead_ctx,
+                    aead,
+                    key_bytes.as_ptr(),
+                    key_bytes.len(),
+                    tag_len,
+                    direction.into(),
+                )
+            },
+            None => unsafe {
+                EVP_AEAD_CTX_init(
+                    *aead_ctx,
+                    aead,
+                    key_bytes.as_ptr(),
+                    key_bytes.len(),
+                    tag_len,
+                    null_mut(),
+                )
+            },
+        } {
+            return Err(Unspecified);
+        }
         Ok(aead_ctx)
     }
+
+    pub fn as_ptr(&self) -> *const EVP_AEAD_CTX {
+        match self {
+            AeadCtx::AES_128_GCM(ctx)
+            | AeadCtx::AES_256_GCM(ctx)
+            | AeadCtx::AES_128_GCM_RANDNONCE(ctx)
+            | AeadCtx::AES_256_GCM_RANDNONCE(ctx)
+            | AeadCtx::AES_128_GCM_TLS12(ctx)
+            | AeadCtx::AES_256_GCM_TLS12(ctx)
+            | AeadCtx::AES_128_GCM_TLS13(ctx)
+            | AeadCtx::AES_256_GCM_TLS13(ctx)
+            | AeadCtx::CHACHA20_POLY1305(ctx) => *ctx.as_const(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn as_mut_ptr(&self) -> *mut EVP_AEAD_CTX {
+        match self {
+            AeadCtx::AES_128_GCM(ctx)
+            | AeadCtx::AES_256_GCM(ctx)
+            | AeadCtx::AES_128_GCM_RANDNONCE(ctx)
+            | AeadCtx::AES_256_GCM_RANDNONCE(ctx)
+            | AeadCtx::AES_128_GCM_TLS12(ctx)
+            | AeadCtx::AES_256_GCM_TLS12(ctx)
+            | AeadCtx::AES_128_GCM_TLS13(ctx)
+            | AeadCtx::AES_256_GCM_TLS13(ctx)
+            | AeadCtx::CHACHA20_POLY1305(ctx) => **ctx,
+        }
```

Comment:
Weren't these changed to `AsRef` and `AsMutRef` impls?

Reply:
I wasn't sure if it would just make it more awkward doing something like `AsRef<LcPtr<*mut EVP_AEAD_CTX>>`, instead of this. Originally it was `AsRef<EVP_AEAD_CTX>` but that was before we migrated this type to the heap. Now that it's being allocated by AWS-LC I'm not sure if we want to / can make assumptions about the alignment of the raw pointer, thus I don't think we should be trying to cast the `*mut EVP_AEAD_CTX` stored in `LcPtr<_>` to a `&EVP_AEAD_CTX`. Anyways I went with the awkward approach for now?